Updated Search page UI

Ananya2001-an opened the pull request:
<!-- If your PR fixes an open issue, use `Closes #999` to link your PR with the issue. #999 stands for the issue number you are fixing -->

## Fixes Issue

closes #3849 

## Changes proposed

<!-- List all the proposed changes in your PR -->
Added filtering option with tags on Search page as well. Users can select and deselect tags to get the desired results based on the combination chosen. For example: if I select tags open source and DevOps then only those users that have both of these tags will be shown. You can click on the tag again to remove the selection. You can also manually type out in the search bar to get the results otherwise use tags directly. When the user selects a tag from the discover page or their LinkFree profile, it will redirect them to the search page showing the results along with the highlighted tag.

<!-- Mark all the applicable boxes. To mark the box as done follow the following conventions -->
<!--
[x] - Correct; marked as done
[X] - Correct; marked as done

[ ] - Not correct; marked as **not** done
-->

## Check List (Check all the applicable boxes) <!-- Follow the above conventions to check the box -->

- [x] My code follows the code style of this project.
- [x] My change requires changes to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] All new and existing tests passed.
- [x] This PR does not contain plagiarized content.
- [x] The title of my pull request is a short description of the requested changes.

## Screenshots

<!-- Add all the screenshots which support your changes -->
![image](https://user-images.githubusercontent.com/55504616/214851736-1e302066-f5f5-4bd3-ac74-bf41bd4eba58.png)

![image](https://user-images.githubusercontent.com/55504616/214851791-37be9a1f-a179-4dda-8ae2-1048da1053ae.png)


## Note to reviewers

<!-- Add notes to reviewers if applicable -->


<a href="https://gitpod.io/#https://github.com/EddieHubCommunity/LinkFree/pull/4030"><img src="https://gitpod.io/button/open-in-gitpod.svg"/></a>

Ananya2001-an commented:
@eddiejaoude @ManishBisht777 Kindly review PR. Made amendments that were causing build fail earlier.

eddiejaoude commented:
Wow this looks great, I will have a play, thank you 👍

Ananya2001-an commented:
Thanks @eddiejaoude . I hope it works out fine.

eddiejaoude commented:
On the search page, when clicking a tag, I get `undefined`

<img width="1390" alt="Screenshot 2023-01-26 at 21 00 34" src="https://user-images.githubusercontent.com/624760/214949687-b6bf04c2-3452-4be4-b909-34251abfe1cd.png">

Ananya2001-an commented:
That's weird. When I tested, it worked fine. I will look into it.

ManishBisht777 commented on components/Tag.js:
```diff
@@ -1,17 +1,26 @@
-        className="flex flex-row p-1 m-2 rounded-lg text-sm font-mono border-2 cursor-pointer shadow-none hover:border-orange-600"
+        href={path && path ==='/search'?
+        (!selected?
+        (search === ''?
```
Here is should be `!search` instead of `search===' '` as on the initial page load the search is `undefined` and you are checking for `""` so it turned out to be false 
hence `/search?search=${search}and${name}` is called but `search is undefined`

that's why the search goes for `undefined and tag`

Ananya2001-an replied:
Yeah I have solved that issue I am just looking at something else now and then I will make a commit

ManishBisht777 reviewed (commented):
Left a comment

ManishBisht777 commented:
[Screencast from 27-01-23 11:35:24 AM IST.webm](https://user-images.githubusercontent.com/89926834/215020660-a6cfe77f-b0f0-4492-9b3d-0be0269857f9.webm)

Rest work great @Ananya2001-an :)

Ananya2001-an commented:
> Screencast.from.27-01-23.11.35.24.AM.IST.webm 
> Rest work great @Ananya2001-an :)

Thanks :)

Ananya2001-an commented:
@eddiejaoude @ManishBisht777 
I have fixed the undefined issue and I also have added more conditions to the href parameter in Tag component since there are cases where user's input in search bar must be in sync with the url search params so that they can use tags along with manual input and get the desired results. This was something that I didn't encounter before but now I have added conditions for it so it would work fine. I have also attached a video below to explain what I am talking about. 


https://user-images.githubusercontent.com/55504616/215046185-477bc902-f47e-494a-bcfc-00e4caf5a39f.mp4

Ananya2001-an commented:
https://user-images.githubusercontent.com/55504616/215047358-e6f9b67d-b4a6-4be7-88e1-2597487bc295.mp4

ManishBisht777 commented on components/Tag.js:
```diff
@@ -1,17 +1,41 @@
       <Link
-        href={`/search?search=${name}`}
-        className="flex flex-row p-1 m-2 rounded-lg text-sm font-mono border-2 cursor-pointer shadow-none hover:border-orange-600"
+        href={path && path ==='/search'?
```
is it possible to extract it in a function rather than doing a large inline condition?. it is hard to know what's going on actually

Ananya2001-an replied:
Ok I will try that once.

Ananya2001-an replied:
Hey @ManishBisht777 I tried using a function but the href prop won't accept it. instead I have created variables for some part of the repetitive conditions so that it's a bit better to understand. It's working as usual.

Ananya2001-an replied:
Also I noticed one problem which I don't understand if it's because of my code or something else but every time I write "ana" in search bar it gives some runtime error. This is not happening with any other user name or language as far as I know. Can u look into it because I am not getting any solution.

https://www.awesomescreenshot.com/video/14344045?key=b86391763fd9f830f3379a6c3d43d1f6

eddiejaoude replied:
Ah that is because one of the users with `ana` in their name is using an unsupported url for images - please do not worry about this, it is being fixed in another PR (it is not any of your changes)

Ananya2001-an replied:
@eddiejaoude Okay that’s a relief. I thought I messed up something 😂

Ananya2001-an commented:
Hey @eddiejaoude can you also review my PR now.

eddiejaoude commented:
I will try to review soon, thank you 👍

eddiejaoude commented:
I had a play and noticed a few things that I think need to be fixed ...
- incomplete words do not search anymore, for example before `open` would find results of `open source` but now it does
- keywords are not matched to the tags (add/removed), for example `python, devrel` are in the search box and are highlighted tags, but when `devrel` is removed it is still highlighted. I think it should work both ways, so if someone searches for `open source` then it matches the keyword as if the tag was clicked

Thank you 👍 This is quite a challenging piece of work

Ananya2001-an commented:
> I had a play and noticed a few things that I think need to be fixed ...
> 
> * incomplete words do not search anymore, for example before `open` would find results of `open source` but now it does
> * keywords are not matched to the tags (add/removed), for example `python, devrel` are in the search box and are highlighted tags, but when `devrel` is removed it is still highlighted. I think it should work both ways, so if someone searches for `open source` then it matches the keyword as if the tag was clicked
> 
> Thank you 👍 This is quite a challenging piece of work

Thanks for the review @eddiejaoude 
The issues that you are talking about I am already aware of it. If I talk about the second point then it was also something that concerned me but the code was already getting so complex that I thought maybe we can just keep it like that because the users are aware that what changes they are making in search bar is getting reflected in the results and it has nothing to do with the selected tags. Only when someone makes any changes in the tags that the search box changes. Basically you can say the search box is the ultimate finder and the tags are just helping us type out those things automatically in it. Besides this I will fix the first issue for sure.

Ananya2001-an commented:
I will still try and fix the second issue as well because obviously that will make the functionality better.

eddiejaoude commented:
Thank you for your quick reply 💯 

- if this can be fixed, so it returns more "greedy" results
- yes let's keep it simple, maybe if they start typing in the search box, it unhighlights all the tags?

Ananya2001-an commented:
Yeah I am on it. If it works then we will have a super cool feature ready to deploy I guess 😎

Ananya2001-an commented:
@eddiejaoude I have made the improvements. Please check and let me know if there's any further issue.

eddiejaoude commented:
Wow I am playing with the latest improvements, these are amazing @Ananya2001-an  🔥

eddiejaoude commented on pages/discover.js:
```diff
@@ -70,18 +72,17 @@ export default function Popular({ data, BASE_URL }) {
               .slice(0, 10)
               .map((tag) => (
-                <Tag name={tag.name} key={tag.name} total={tag.total} />
+                <Tag name={tag.name} key={tag.name} total={tag.total} path='/discover'/>
```
I get an error when clicking on a tag from the Discover page

<img width="1680" alt="Screenshot 2023-02-01 at 15 15 52" src="https://user-images.githubusercontent.com/624760/215987909-f61019e8-026c-43ef-9ab4-51525ced24fe.png">

Ananya2001-an replied:
Yeah I will write the condition for undefined

Ananya2001-an replied:
If there are any other issues that you noticed do let me know

eddiejaoude replied:
I think this is the last one, thank you for your persistence 💪 hopefully we will get this merged very soon 🎉

Ananya2001-an replied:
Great! I will test everything one last time and make the commit soon.

Ananya2001-an replied:
@eddiejaoude made the last commit with all fixes needed.

Ananya2001-an commented:
> Wow I am playing with the latest improvements, these are amazing @Ananya2001-an  🔥 

Thank you 😊

ManishBisht777 commented:
@Ananya2001-an sorry for not being active on the pr ( GitHub actually), I have been facing many issues with my internet connection since 1 week so I had to change the provider.

Great work @Ananya2001-an the changes looks amazing :fire: :fire: 

The small addons are a cherry on top
![image](https://user-images.githubusercontent.com/89926834/216113293-cb0dddd8-e4be-48da-aacc-87fad2b3ea6d.png)

Ananya2001-an commented:
> @Ananya2001-an sorry for not being active on the pr ( GitHub actually), I have been facing many issues with my internet connection since 1 week so I had to change the provider.
> 
> Great work @Ananya2001-an the changes looks amazing :fire: :fire: 
> 
> The small addons are a cherry on top
> ![image](https://user-images.githubusercontent.com/89926834/216113293-cb0dddd8-e4be-48da-aacc-87fad2b3ea6d.png)
> 

No worries. Thanks for your help till now. Really appreciate it. 😊

Ananya2001-an commented:
@eddiejaoude I have fixed the undefined error.

eddiejaoude commented:
thanks @Ananya2001-an 👍 

I still get the error, let me merge to a temporary branch and format the code to our coding standards and see if I can take a deeper look - you have done all the hard work already

Ananya2001-an commented:
@eddiejaoude Thanks Eddie. If you are still getting any issue please feel free to tell me.

eddiejaoude commented:
sure, thank you @Ananya2001-an ❤️

eddiejaoude commented on components/Tag.js:
```diff
@@ -1,17 +1,33 @@
       <Link
-        href={`/search?search=${name}`}
-        className="flex flex-row p-1 m-2 rounded-lg text-sm font-mono border-2 cursor-pointer shadow-none hover:border-orange-600"
+        onClick={()=>{
```
I added ...

```js
          if (!currentInput) {
            return;
          }
```

and I think it fixed it, let me push to a branch @Ananya2001-an and you can review if you have time

Ananya2001-an replied:
Yes sure I will check

eddiejaoude replied:
I pushed to here https://github.com/EddieHubCommunity/LinkFree/pull/4375
